### PR TITLE
MAISTRA-1632 add conversion webhook to SMCP resource

### PR DESCRIPTION
**Pull request by rcernich**
Signed-off-by: rcernich <rcernich@redhat.com>

### CODE REVIEW

File: pkg/controller/servicemesh/webhookca/conversion.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+	logger := common.LogFromContext(ctx)
+	if mw.Spec.Conversion == nil || mw.Spec.Conversion.Strategy != apixv1.WebhookConverter || mw.Spec.Conversion.Webhook == nil || mw.Spec.Conversion.Webhook.ClientConfig == nil {
+		logger.Info("Skipping CABundle update, no webhook client.")
+		return nil
+	}
+	if bytes.Compare(mw.Spec.Conversion.Webhook.ClientConfig.CABundle, caBundle) == 0 {
+		logger.Info("Correct CABundle already present. Ignoring")
+		return nil
+	}
+	logger.Info("Udating CABundle")
```
Comment (dgn) [inline pkg/controller/servicemesh/webhookca/conversion.go]: nit: missing a `p`

File: pkg/controller/servicemesh/webhooks/hacks.go [hunk 2]
```diff
@@ -107,6 +111,47 @@ func createWebhookResources(ctx context.Context, mgr manager.Manager, log logr.L
 		return err
 	}
 
+	log.Info("Adding conversion webhook to SCMP CRD")
```
Comment (dgn) [inline pkg/controller/servicemesh/webhooks/hacks.go]: nit: `SMCP`

File: pkg/controller/servicemesh/webhookca/wrappers.go [hunk 3]
```diff
@@ -18,17 +21,42 @@ type webhookGetter interface {
...
+func updateAdmissionWebHookCABundles(ctx context.Context, cl client.Client, currentConfig webhookWrapper, caBundle []byte) error {
+	logger := common.LogFromContext(ctx)
+	updated := false
+	newConfig := currentConfig.Copy().(webhookWrapper)
+	for _, clientConfig := range newConfig.ClientConfigs() {
+		updated = common.InjectCABundle(clientConfig, caBundle) || updated
+	}
+
+	if updated {
+		logger.Info("Udating CABundle")
```
Comment (dgn) [inline pkg/controller/servicemesh/webhookca/wrappers.go]: nit: missing `p`

File: pkg/controller/servicemesh/webhookca/wrappers.go [hunk 4]
```diff
@@ -18,17 +21,42 @@ type webhookGetter interface {
...
+	case *v1.ValidatingWebhookConfiguration:
 		return &validatingWebhookWrapper{ValidatingWebhookConfiguration: wh}, nil
-	case *v1beta1.MutatingWebhookConfiguration:
+	case *v1.MutatingWebhookConfiguration:
 		return &mutatingWebhookWrapper{MutatingWebhookConfiguration: wh}, nil
+	case *apixv1.CustomResourceDefinition:
+		return &conversionWebhookWrapper{CustomResourceDefinition: wh}, nil
 	}
-	return nil, fmt.Errorf("Object is not a MutatingWebhookConfiguration or ValidatingWebhookConfiguration")
+	return nil, fmt.Errorf("Object is not a MutatingWebhookConfiguration or ValidatingWebhookConfiguration: %T", obj)
```
Comment (dgn) [inline pkg/controller/servicemesh/webhookca/wrappers.go]: ...and it's also not a CRD - I guess now we should use the `object not in [list]` style

File: pkg/apis/maistra/v1/conversion.go [hunk 5]
```diff
@@ -0,0 +1,23 @@
+package v1
+
+import (
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/conversion"
+)
+
+var _ conversion.Convertible = (*ServiceMeshControlPlane)(nil)
```
Comment (dgn) [inline pkg/apis/maistra/v1/conversion.go:8]: :car:

**Review by dgn — APPROVED** [review]
lgtm, just a few typos (most of which were already there before I think)

File: pkg/controller/servicemesh/webhookca/wrappers.go (hunk 4 above)
Comment (rcernich) [reply pkg/controller/servicemesh/webhookca/wrappers.go]: done

File: pkg/controller/servicemesh/webhookca/wrappers.go (hunk 3 above)
Comment (rcernich) [reply pkg/controller/servicemesh/webhookca/wrappers.go]: done

File: pkg/controller/servicemesh/webhooks/hacks.go (hunk 2 above)
Comment (rcernich) [reply pkg/controller/servicemesh/webhooks/hacks.go]: done

File: pkg/controller/servicemesh/webhookca/conversion.go (hunk 1 above)
Comment (rcernich) [reply pkg/controller/servicemesh/webhookca/conversion.go]: done

**Comment by rcernich** [comment]
Latest updates are the result of testing with v1 resources.

**Review by jwendell — APPROVED** [review]

**Pull request merged**